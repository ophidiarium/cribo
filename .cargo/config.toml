[build]
rustflags = ["-Ctarget-cpu=native"]

[alias]
# Coverage commands using cargo-llvm-cov
coverage = "llvm-cov --html --open"
coverage-text = "llvm-cov --text"
coverage-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov.info"
coverage-json = "llvm-cov --json --output-path target/llvm-cov/coverage.json"
coverage-clean = "llvm-cov clean"

# Branch coverage commands (requires nightly Rust)
coverage-branch = "llvm-cov --html --open --branch"
coverage-branch-text = "llvm-cov --text --branch"
coverage-branch-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov-branch.info --branch"
