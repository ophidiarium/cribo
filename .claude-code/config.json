{
    "hooks": {
        "preToolUse": "./.claude-code/hooks/prevent-main-commit.sh"
    }
}
