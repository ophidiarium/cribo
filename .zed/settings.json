{
    "file_scan_exclusions": [
        "**/.git",
        "**/Thumbs.dn",
        "**/.vscode",
        "**/__pycache__",
        "**/.mypy_cache",
        "**/.pytest_cache",
        "**/.ruff_cache",
        "**/.ropeproject"
    ],
    "auto_install_extensions": {
        "toml": true,
        "typos": true,
        "ruff": true,
        "mcp-server-github": true,
        "mcp-server-context7": true
    },
    "terminal": {
        "detect_venv": {
            "on": {
                "directories": [".venv"]
            }
        }
    }
}
