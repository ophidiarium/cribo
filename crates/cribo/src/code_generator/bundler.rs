use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use indexmap::{IndexMap as FxIndexMap, IndexSet as FxIndexSet};
use log::{debug, trace, warn};
use ruff_python_ast::{
    self as ast, Arguments, CmpOp, Expr, ExprAttribute, ExprCall, ExprCompare, ExprContext,
    ExprDict, ExprList, ExprName, ExprStringLiteral, ExprSubscript, ExprTuple, Identifier, Int,
    Keyword, ModModule, Stmt, StmtAnnAssign, StmtAssign, StmtClassDef, StmtExpr, StmtFunctionDef,
    StmtIf, StmtImport, StmtImportFrom, StmtReturn, StmtWith, WithItem, visitor::Visitor,
};
use ruff_text_size::{Ranged, TextRange, TextSize};
use rustc_hash::{FxHashMap, FxHashSet};

use crate::{
    code_generator::{
        circular_deps::SymbolDependencyGraph,
        context::{
            BundleParams, DirectImportContext, HardDependency, InlineContext,
            ModuleTransformContext, ProcessGlobalsParams, SemanticContext, TransformFunctionParams,
        },
        globals::{GlobalsLifter, transform_globals_in_expr, transform_globals_in_stmt},
        import_transformer::{RecursiveImportTransformer, RecursiveImportTransformerParams},
    },
    cribo_graph::CriboGraph as DependencyGraph,
    semantic_bundler::{ModuleGlobalInfo, SemanticBundler, SymbolRegistry},
    transformation_context::TransformationContext,
    visitors::{ImportDiscoveryVisitor, NoOpsRemovalTransformer},
};

/// This approach avoids forward reference issues while maintaining Python module semantics
#[derive(Debug)]
pub struct HybridStaticBundler<'a> {
    /// Track if importlib was fully transformed and should be removed
    pub(crate) importlib_fully_transformed: bool,
    /// Map from original module name to synthetic module name
    pub(crate) module_registry: FxIndexMap<String, String>,
    /// Map from synthetic module name to init function name
    pub(crate) init_functions: FxIndexMap<String, String>,
    /// Collected future imports
    pub(crate) future_imports: FxIndexSet<String>,
    /// Collected stdlib imports that are safe to hoist
    /// Maps module name to map of imported names to their aliases (None if no alias)
    pub(crate) stdlib_import_from_map: FxIndexMap<String, FxIndexMap<String, Option<String>>>,
    /// Regular import statements (import module)
    pub(crate) stdlib_import_statements: Vec<Stmt>,
    /// Track which modules have been bundled
    pub(crate) bundled_modules: FxIndexSet<String>,
    /// Modules that were inlined (not wrapper modules)
    pub(crate) inlined_modules: FxIndexSet<String>,
    /// Entry point path for calculating relative paths
    pub(crate) entry_path: Option<String>,
    /// Entry module name
    pub(crate) entry_module_name: String,
    /// Whether the entry is __init__.py or __main__.py
    pub(crate) entry_is_package_init_or_main: bool,
    /// Module export information (for __all__ handling)
    pub(crate) module_exports: FxIndexMap<String, Option<Vec<String>>>,
    /// Lifted global declarations to add at module top level
    pub(crate) lifted_global_declarations: Vec<Stmt>,
    /// Modules that are imported as namespaces (e.g., from package import module)
    /// Maps module name to set of importing modules
    pub(crate) namespace_imported_modules: FxIndexMap<String, FxIndexSet<String>>,
    /// Reference to the central module registry
    pub(crate) module_info_registry: Option<&'a crate::orchestrator::ModuleRegistry>,
    /// Modules that are part of circular dependencies
    pub(crate) circular_modules: FxIndexSet<String>,
    /// Pre-declared symbols for circular modules (module -> symbol -> renamed)
    pub(crate) circular_predeclarations: FxIndexMap<String, FxIndexMap<String, String>>,
    /// Hard dependencies that need to be hoisted
    pub(crate) hard_dependencies: Vec<HardDependency>,
    /// Symbol dependency graph for circular modules
    pub(crate) symbol_dep_graph: SymbolDependencyGraph,
    /// Module ASTs for resolving re-exports
    pub(crate) module_asts: Option<Vec<(String, ModModule, PathBuf, String)>>,
    /// Global registry of deferred imports to prevent duplication
    /// Maps (module_name, symbol_name) to the source module that deferred it
    pub(crate) global_deferred_imports: FxIndexMap<(String, String), String>,
    /// Track all namespaces that need to be created before module initialization
    /// This ensures parent namespaces exist before any submodule assignments
    pub(crate) required_namespaces: FxIndexSet<String>,
    /// Runtime tracking of all created namespaces to prevent duplicates
    /// This includes both pre-identified and dynamically created namespaces
    pub(crate) created_namespaces: FxIndexSet<String>,
    /// Modules that have explicit __all__ defined
    pub(crate) modules_with_explicit_all: FxIndexSet<String>,
    /// Transformation context for tracking node mappings
    pub(crate) transformation_context: TransformationContext,
    /// Module/symbol pairs that should be kept after tree shaking
    pub(crate) tree_shaking_keep_symbols: Option<indexmap::IndexSet<(String, String)>>,
    /// Whether to use the module cache model for circular dependencies
    pub(crate) use_module_cache_model: bool,
}

impl<'a> Default for HybridStaticBundler<'a> {
    fn default() -> Self {
        Self {
            importlib_fully_transformed: false,
            module_registry: FxIndexMap::default(),
            init_functions: FxIndexMap::default(),
            future_imports: FxIndexSet::default(),
            stdlib_import_from_map: FxIndexMap::default(),
            stdlib_import_statements: Vec::new(),
            bundled_modules: FxIndexSet::default(),
            inlined_modules: FxIndexSet::default(),
            entry_path: None,
            entry_module_name: String::new(),
            entry_is_package_init_or_main: false,
            module_exports: FxIndexMap::default(),
            lifted_global_declarations: Vec::new(),
            namespace_imported_modules: FxIndexMap::default(),
            module_info_registry: None,
            circular_modules: FxIndexSet::default(),
            circular_predeclarations: FxIndexMap::default(),
            hard_dependencies: Vec::new(),
            symbol_dep_graph: SymbolDependencyGraph::default(),
            module_asts: None,
            global_deferred_imports: FxIndexMap::default(),
            required_namespaces: FxIndexSet::default(),
            created_namespaces: FxIndexSet::default(),
            modules_with_explicit_all: FxIndexSet::default(),
            transformation_context: TransformationContext::new(),
            tree_shaking_keep_symbols: None,
            use_module_cache_model: true,
        }
    }
}

// Main implementation
impl<'a> HybridStaticBundler<'a> {
    /// Create a new bundler instance
    pub fn new() -> Self {
        Self::default()
    }

    /// Check if a statement is a hoisted import
    pub fn is_hoisted_import(&self, stmt: &Stmt) -> bool {
        match stmt {
            Stmt::Import(import) => {
                // Check if any alias is in our stdlib imports
                import.names.iter().any(|alias| {
                    self.stdlib_import_from_map
                        .contains_key(&alias.name.to_string())
                })
            }
            Stmt::ImportFrom(from_import) => {
                if let Some(module) = &from_import.module {
                    let module_name = module.to_string();
                    self.stdlib_import_from_map.contains_key(&module_name)
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    /// Resolve a relative import with context
    pub fn resolve_relative_import_with_context(
        &self,
        current_module: &str,
        module_path: &Path,
        level: u32,
        module: Option<&str>,
    ) -> Option<String> {
        // Implementation from original file
        // This is a placeholder - the actual implementation is quite complex
        None
    }

    /// Filter exports based on tree shaking
    pub fn filter_exports_by_tree_shaking(
        &self,
        keep_symbols: Option<&indexmap::IndexSet<(String, String)>>,
        module_name: &str,
        exports: &[String],
    ) -> Vec<String> {
        if let Some(keep_symbols) = keep_symbols {
            exports
                .iter()
                .filter(|symbol| {
                    keep_symbols.contains(&(module_name.to_string(), symbol.to_string()))
                })
                .cloned()
                .collect()
        } else {
            exports.to_vec()
        }
    }

    // More methods to be moved from the original implementation...
}

/// Main entry point for bundling modules
pub fn bundle_modules(params: BundleParams) -> Result<ModModule> {
    // This function implementation is very large and should be moved from the original file
    // For now, creating a placeholder
    let mut bundler = HybridStaticBundler::new();

    // The actual implementation involves:
    // 1. Processing modules and collecting information
    // 2. Handling circular dependencies
    // 3. Generating the bundled output
    // 4. Managing imports and exports

    // Placeholder return
    Ok(ModModule {
        range: TextRange::default(),
        body: vec![],
        node_index: Default::default(),
    })
}

// Additional implementation blocks and helper functions should be moved here
// from the original code_generator.rs file
