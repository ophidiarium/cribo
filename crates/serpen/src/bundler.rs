use anyhow::{anyhow, Context, Result};
use log::{debug, info, warn};
use rustpython_parser::ast::{Mod, Stmt, StmtImportFrom};
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use crate::config::Config;
use crate::dependency_graph::{DependencyGraph, ModuleNode};
use crate::emit::CodeEmitter;
use crate::resolver::{ImportType, ModuleResolver};

/// Type alias for module processing queue
type ModuleQueue = Vec<(String, PathBuf)>;
/// Type alias for processed modules set
type ProcessedModules = HashSet<String>;

/// Parameters for discovery phase operations
struct DiscoveryParams<'a> {
  resolver: &'a mut ModuleResolver,
  modules_to_process: &'a mut ModuleQueue,
  processed_modules: &'a ProcessedModules,
  queued_modules: &'a mut HashSet<String>,
}

pub struct Bundler {
  config: Config,
}

impl Bundler {
  pub fn new(config: Config) -> Self {
    Self { config }
  }

  /// Main bundling function
  pub fn bundle(
    &mut self,
    entry_path: &Path,
    output_path: &Path,
    emit_requirements: bool,
  ) -> Result<()> {
    info!("Starting bundle process");
    debug!("Entry: {:?}, Output: {:?}", entry_path, output_path);

    // Auto-detect the entry point's directory as a source directory
    if let Some(entry_dir) = entry_path.parent() {
      // Canonicalize the path to avoid duplicates due to different lexical representations
      let entry_dir = match entry_dir.canonicalize() {
        Ok(canonical_path) => canonical_path,
        Err(_) => {
          // Fall back to the original path if canonicalization fails (e.g., path doesn't exist)
          entry_dir.to_path_buf()
        }
      };
      if !self.config.src.contains(&entry_dir) {
        debug!("Adding entry directory to src paths: {:?}", entry_dir);
        self.config.src.insert(0, entry_dir);
      }
    }

    // Initialize resolver
    let mut resolver = ModuleResolver::new(self.config.clone())?;

    // Find the entry module name
    let entry_module_name = self.find_entry_module_name(entry_path, &resolver)?;
    info!("Entry module: {}", entry_module_name);

    // Build dependency graph
    let mut graph = self.build_dependency_graph(entry_path, &entry_module_name, &mut resolver)?;

    // Filter to only modules reachable from entry
    debug!(
      "Before filtering - graph has {} modules",
      graph.get_modules().len()
    );
    graph = graph.filter_reachable_from(&entry_module_name)?;
    debug!(
      "After filtering - graph has {} modules",
      graph.get_modules().len()
    );

    // Check for cycles
    if graph.has_cycles() {
      return Err(anyhow!(
        "Circular dependencies detected in the module graph"
      ));
    }

    // Get topologically sorted modules
    let sorted_modules = graph.topological_sort()?;
    info!("Found {} modules to bundle", sorted_modules.len());
    for (i, module) in sorted_modules.iter().enumerate() {
      debug!("Module {}: {} ({:?})", i, module.name, module.path);
    }

    // Generate bundled code
    let mut emitter = CodeEmitter::new(
      resolver,
      self.config.preserve_comments,
      self.config.preserve_type_hints,
    );

    let bundled_code = emitter.emit_bundle(&sorted_modules, &entry_module_name)?;

    // Write output file
    fs::write(output_path, bundled_code)
      .with_context(|| format!("Failed to write output file: {:?}", output_path))?;

    info!("Bundle written to: {:?}", output_path);

    // Generate requirements.txt if requested
    if emit_requirements {
      self.write_requirements_file(&sorted_modules, &mut emitter, output_path)?;
    }

    Ok(())
  }

  /// Helper method to find module name in source directories
  fn find_module_in_src_dirs(&self, entry_path: &Path) -> Option<String> {
    for src_dir in &self.config.src {
      let relative_path = entry_path.strip_prefix(src_dir).ok()?;
      if let Some(module_name) = self.path_to_module_name(relative_path) {
        return Some(module_name);
      }
    }
    None
  }

  /// Find the module name for the entry script
  fn find_entry_module_name(
    &self,
    entry_path: &Path,
    _resolver: &ModuleResolver,
  ) -> Result<String> {
    // Try to find which src directory contains the entry file
    if let Some(module_name) = self.find_module_in_src_dirs(entry_path) {
      return Ok(module_name);
    }

    // If not found in src directories, use the file stem as module name
    let module_name = entry_path
      .file_stem()
      .and_then(|name| name.to_str())
      .ok_or_else(|| {
        anyhow!(
          "Cannot determine module name from entry path: {:?}",
          entry_path
        )
      })?;

    Ok(module_name.to_string())
  }

  /// Convert a relative path to a module name
  fn path_to_module_name(&self, relative_path: &Path) -> Option<String> {
    let parts: Vec<String> = relative_path
      .components()
      .map(|c| c.as_os_str().to_string_lossy().to_string())
      .collect();

    if parts.is_empty() {
      return None;
    }

    let mut module_parts = parts;
    let last_part = module_parts.last_mut()?;

    // Remove .py extension
    if last_part.ends_with(".py") {
      *last_part = last_part[..last_part.len() - 3].to_string();
    }

    // Handle __init__.py
    if last_part == "__init__" {
      module_parts.pop();
    }

    if module_parts.is_empty() {
      return None;
    }

    Some(module_parts.join("."))
  }

  /// Build the complete dependency graph starting from the entry module
  fn build_dependency_graph(
    &self,
    entry_path: &Path,
    entry_module_name: &str,
    resolver: &mut ModuleResolver,
  ) -> Result<DependencyGraph> {
    let mut graph = DependencyGraph::new();
    let mut processed_modules = ProcessedModules::new();
    let mut queued_modules = HashSet::new();
    let mut modules_to_process = ModuleQueue::new();
    modules_to_process.push((entry_module_name.to_string(), entry_path.to_path_buf()));
    queued_modules.insert(entry_module_name.to_string());

    // Store module data for phase 2
    type ModuleData = (String, PathBuf, Vec<String>);
    let mut all_modules: Vec<ModuleData> = Vec::new();

    // PHASE 1: Discover and collect all modules
    info!("Phase 1: Discovering all modules...");
    while let Some((module_name, module_path)) = modules_to_process.pop() {
      debug!("Discovering module: {} ({:?})", module_name, module_path);
      if processed_modules.contains(&module_name) {
        debug!("Module {} already discovered, skipping", module_name);
        continue;
      }

      // Parse the module and extract imports
      let imports = self.extract_imports(&module_path)?;
      debug!("Extracted imports from {}: {:?}", module_name, imports);

      // Store module data for later processing
      all_modules.push((module_name.clone(), module_path.clone(), imports.clone()));
      processed_modules.insert(module_name.clone());

      // Find and queue first-party imports for discovery
      for import in imports {
        let mut params = DiscoveryParams {
          resolver,
          modules_to_process: &mut modules_to_process,
          processed_modules: &processed_modules,
          queued_modules: &mut queued_modules,
        };
        self.process_import_for_discovery(&import, &mut params);
      }
    }

    info!("Phase 1 complete: discovered {} modules", all_modules.len());

    // PHASE 2: Add all modules to graph and create dependency edges
    info!("Phase 2: Adding modules to graph...");

    // First, add all modules to the graph
    for (module_name, module_path, imports) in &all_modules {
      let module_node = ModuleNode {
        name: module_name.clone(),
        path: module_path.clone(),
        imports: imports.clone(),
      };
      debug!("Adding module to graph: {}", module_node.name);
      graph.add_module(module_node);
    }

    info!("Added {} modules to graph", graph.get_modules().len());

    // Then, add all dependency edges
    info!("Phase 2: Creating dependency edges...");
    for (module_name, _module_path, imports) in &all_modules {
      for import in imports {
        self.process_import_for_dependency(import, module_name, (resolver, &mut graph));
      }
    }

    info!(
      "Phase 2 complete: dependency graph built with {} modules",
      graph.get_modules().len()
    );
    Ok(graph)
  }

  /// Extract import statements from a Python file using AST parsing
  /// This handles all import variations including multi-line, aliased, relative, and parenthesized imports
  pub fn extract_imports(&self, file_path: &Path) -> Result<Vec<String>> {
    let source = fs::read_to_string(file_path)
      .with_context(|| format!("Failed to read file: {:?}", file_path))?;

    let parsed = rustpython_parser::parse(
      &source,
      rustpython_parser::Mode::Module,
      file_path.to_string_lossy().as_ref(),
    )
    .with_context(|| format!("Failed to parse Python file: {:?}", file_path))?;

    let mut imports = Vec::new();

    if let Mod::Module(module) = parsed {
      for stmt in module.body.iter() {
        self.extract_imports_from_statement(stmt, &mut imports);
      }
    }

    Ok(imports)
  }

  /// Extract import module names from a single AST statement
  fn extract_imports_from_statement(&self, stmt: &Stmt, imports: &mut Vec<String>) {
    if let Stmt::Import(import_stmt) = stmt {
      for alias in &import_stmt.names {
        imports.push(alias.name.to_string());
      }
    } else if let Stmt::ImportFrom(import_from_stmt) = stmt {
      self.process_import_from_statement(import_from_stmt, imports);
    }
  }

  /// Process a "from ... import ..." statement to extract module names
  fn process_import_from_statement(
    &self,
    import_from_stmt: &StmtImportFrom,
    imports: &mut Vec<String>,
  ) {
    if let Some(ref module) = import_from_stmt.module {
      let level = import_from_stmt.level.map(|i| i.to_u32()).unwrap_or(0);
      let module_name = self.format_module_name(module, level);
      imports.push(module_name);
    } else if let Some(level_int) = import_from_stmt.level {
      let level = level_int.to_u32();
      if level > 0 {
        imports.push(".".repeat(level as usize));
      }
    }
  }

  /// Format module name based on relative import level
  fn format_module_name(&self, module: &str, level: u32) -> String {
    if level > 0 {
      if module.is_empty() {
        ".".to_string()
      } else {
        let dots = ".".repeat(level as usize);
        format!("{}{}", dots, module)
      }
    } else {
      module.to_string()
    }
  }

  /// Helper method to add module to discovery queue if not already processed or queued
  fn add_to_discovery_queue_if_new(
    &self,
    import: &str,
    import_path: PathBuf,
    discovery_params: &mut DiscoveryParams,
  ) {
    if !discovery_params.processed_modules.contains(import)
      && !discovery_params.queued_modules.contains(import)
    {
      debug!("Adding '{}' to discovery queue", import);
      discovery_params
        .modules_to_process
        .push((import.to_string(), import_path));
      discovery_params.queued_modules.insert(import.to_string());
    } else {
      debug!("Module '{}' already processed or queued, skipping", import);
    }
  }

  /// Process an import during discovery phase
  fn process_import_for_discovery(&self, import: &str, params: &mut DiscoveryParams) {
    match params.resolver.classify_import(import) {
      ImportType::FirstParty => {
        debug!("'{}' classified as FirstParty", import);
        if let Ok(Some(import_path)) = params.resolver.resolve_module_path(import) {
          debug!("Resolved '{}' to path: {:?}", import, import_path);
          self.add_to_discovery_queue_if_new(import, import_path, params);
        } else {
          warn!("Failed to resolve path for first-party module: {}", import);
        }
      }
      ImportType::ThirdParty | ImportType::StandardLibrary => {
        debug!("'{}' classified as external (preserving)", import);
      }
    }
  }

  /// Process an import during dependency graph creation phase
  fn process_import_for_dependency(
    &self,
    import: &str,
    module_name: &str,
    resolver_and_graph: (&ModuleResolver, &mut DependencyGraph),
  ) {
    let (resolver, graph) = resolver_and_graph;
    match resolver.classify_import(import) {
      ImportType::FirstParty => {
        // Add dependency edge (dependency -> dependent)
        // This means the imported module must come before the importing module
        debug!("Adding dependency edge: {} -> {}", import, module_name);
        if let Err(e) = graph.add_dependency(import, module_name) {
          debug!("Failed to add dependency edge: {}", e);
          warn!("Failed to add dependency edge: {}", e);
        } else {
          debug!(
            "Successfully added dependency edge: {} -> {}",
            import, module_name
          );
        }
      }
      ImportType::ThirdParty | ImportType::StandardLibrary => {
        // These will be preserved in the output, not inlined
      }
    }
  }

  /// Write requirements.txt file if there are dependencies
  fn write_requirements_file(
    &self,
    sorted_modules: &[&ModuleNode],
    emitter: &mut CodeEmitter,
    output_path: &Path,
  ) -> Result<()> {
    let requirements_content = emitter.generate_requirements(sorted_modules)?;
    if !requirements_content.is_empty() {
      let requirements_path = output_path
        .parent()
        .unwrap_or_else(|| Path::new("."))
        .join("requirements.txt");

      fs::write(&requirements_path, requirements_content)
        .with_context(|| format!("Failed to write requirements file: {:?}", requirements_path))?;

      info!("Requirements written to: {:?}", requirements_path);
    } else {
      info!("No third-party dependencies found, skipping requirements.txt");
    }
    Ok(())
  }
}
