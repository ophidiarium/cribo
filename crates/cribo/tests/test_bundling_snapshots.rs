#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use anyhow::Result;
use std::fs;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

use cribo::bundler::Bundler;
use cribo::config::Config;
use cribo::util::get_python_executable;

// Ruff linting integration for cross-validation
use ruff_linter::linter::{ParseSource, lint_only};
use ruff_linter::registry::Rule;
use ruff_linter::settings::{LinterSettings, flags};
use ruff_linter::source_kind::SourceKind;
use ruff_python_ast::PySourceType;

/// Structured execution results for better snapshot formatting
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct ExecutionResults {
    status: ExecutionStatus,
    stdout: String,
    stderr: String,
}

/// Sanitize file paths in error messages to make them deterministic for snapshots
fn sanitize_paths(text: &str) -> String {
    use regex::Regex;

    // Match common temporary directory patterns and replace with a generic placeholder
    let patterns = vec![
        // Unix/macOS temp paths like /var/folders/xyz/abc123/T/.tmpXYZ/file.py
        (r"/var/folders/[^/]+/[^/]+/T/\.[^/]+/", "<TMP>/"),
        // Standard Unix temp paths like /tmp/.tmpXYZ/file.py
        (r"/tmp/\.[^/]+/", "<TMP>/"),
        // Windows temp paths
        (
            r"C:\\Users\\[^\\]+\\AppData\\Local\\Temp\\[^\\]+\\",
            "<TMP>/",
        ),
        // Generic temp directory patterns
        (r"/[Tt]emp/[^/]+/", "<TMP>/"),
        // Python installation paths (varying versions and locations)
        (
            r"/opt/homebrew/Cellar/python@[\d.]+/[\d._]+/Frameworks/Python\.framework/Versions/[\d.]+/lib/python[\d.]+/",
            "<PYTHON_LIB>/",
        ),
        (r"/usr/lib/python[\d.]+/", "<PYTHON_LIB>/"),
        (r"C:\\Python\d+\\lib\\", "<PYTHON_LIB>/"),
        // Windows hosted tool cache paths (GitHub Actions)
        (
            r"C:\\hostedtoolcache\\windows\\Python\\[\d.]+\\x64\\Lib\\",
            "<PYTHON_LIB>/",
        ),
    ];

    let mut result = text.to_string();
    for (pattern, replacement) in patterns {
        let re = Regex::new(pattern).expect("Invalid regex pattern");
        result = re.replace_all(&result, replacement).to_string();
    }

    // Normalize Python error formatting differences between versions
    // Replace line numbers in importlib which vary between Python versions
    let importlib_line_re = Regex::new(r"line \d+, in import_module").expect("Invalid regex");
    result = importlib_line_re
        .replace_all(&result, "line <LINE>, in import_module")
        .to_string();

    // Remove lines that only contain caret/tilde indicators as they vary between Python versions
    let indicator_line_re = Regex::new(r"(?m)^\s+[\^~]+\s*\n").expect("Invalid regex");
    result = indicator_line_re.replace_all(&result, "").to_string();

    // Normalize Windows path separators in sanitized paths
    result = result.replace("<PYTHON_LIB>/importlib\\", "<PYTHON_LIB>/importlib/");

    result
}

#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
enum ExecutionStatus {
    Success,
    Failed(i32),
}

/// Ruff linting results for cross-validation
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct RuffLintResults {
    f401_violations: Vec<String>, // Unused imports
    f404_violations: Vec<String>, // Late future imports
    other_violations: Vec<String>,
    total_violations: usize,
}

/// Run ruff linting on bundled code to cross-validate import handling
fn run_ruff_lint_on_bundle(bundled_code: &str) -> RuffLintResults {
    // Create settings for multiple import-related rules with both F401 and F404 enabled
    let settings = LinterSettings {
        rules: [Rule::UnusedImport, Rule::LateFutureImport]
            .into_iter()
            .collect(),
        ..LinterSettings::default()
    };

    let path = Path::new("<bundled>.py");
    let source_kind = SourceKind::Python(bundled_code.to_string());

    let result = lint_only(
        path,
        None,
        &settings,
        flags::Noqa::Enabled,
        &source_kind,
        PySourceType::Python,
        ParseSource::None,
    );

    let mut f401_violations = Vec::new();
    let mut f404_violations = Vec::new();
    let mut other_violations = Vec::new();

    for message in &result.messages {
        if let Some(rule) = message.to_rule() {
            let location = message.compute_start_location();
            let violation_info = format!(
                "Line {}: {} - {}",
                location.line.get(),
                rule.noqa_code(),
                message.body()
            );

            match rule {
                Rule::UnusedImport => f401_violations.push(violation_info),
                Rule::LateFutureImport => f404_violations.push(violation_info),
                _ => other_violations.push(violation_info),
            }
        }
    }

    let total_violations = f401_violations.len() + f404_violations.len() + other_violations.len();

    RuffLintResults {
        f401_violations,
        f404_violations,
        other_violations,
        total_violations,
    }
}

/// Generic test that processes all fixture directories in tests/fixtures/bundling
/// Each directory should contain a main.py entry point and will be bundled and executed
#[test]
fn test_all_bundling_fixtures() -> Result<()> {
    let fixtures_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("fixtures")
        .join("bundling");

    // Skip if bundling directory doesn't exist
    if !fixtures_dir.exists() {
        return Ok(());
    }

    // Read all subdirectories in the bundling fixtures directory
    let entries = fs::read_dir(&fixtures_dir)?;
    let mut fixture_names = Vec::new();

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_dir() {
            if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                fixture_names.push(name.to_string());
            }
        }
    }

    // Sort for deterministic test order
    fixture_names.sort();

    // Process each fixture directory
    for fixture_name in fixture_names {
        test_single_bundling_fixture(&fixtures_dir, &fixture_name)?;
    }

    Ok(())
}

/// Test a single bundling fixture directory
fn test_single_bundling_fixture(fixtures_dir: &Path, fixture_name: &str) -> Result<()> {
    let fixture_path = fixtures_dir.join(fixture_name);
    let main_py_path = fixture_path.join("main.py");

    // Skip if main.py doesn't exist
    if !main_py_path.exists() {
        // Skip silently or use conditional verbose logging
        if std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!("Skipping {}: no main.py found", fixture_name);
        }
        return Ok(());
    }

    // Check if this is an expected failure fixture (xfail prefix)
    let expects_failure = fixture_name.starts_with("xfail_");

    // Create temporary directory for output
    let temp_dir = TempDir::new()?;
    let bundle_path = temp_dir.path().join("bundled.py");

    // Configure bundler
    let config = Config::default();
    let mut bundler = Bundler::new(config);

    // Bundle the fixture
    bundler.bundle(&main_py_path, &bundle_path, false)?;

    // Optionally validate Python syntax before execution
    let python_cmd = get_python_executable();
    let syntax_check = Command::new(&python_cmd)
        .args(["-m", "py_compile"])
        .arg(&bundle_path)
        .output();
    if let Ok(output) = syntax_check {
        if !output.status.success() && std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!(
                "Warning: Bundled code has syntax errors for fixture {}",
                fixture_name
            );
            eprintln!("Stderr: {}", String::from_utf8_lossy(&output.stderr));
        }
    }

    // Read the bundled code and normalize line endings for cross-platform compatibility
    let bundled_code = fs::read_to_string(&bundle_path)?;

    // Run ruff linting for cross-validation of unused imports elimination
    let ruff_results = run_ruff_lint_on_bundle(&bundled_code);

    // Execute the bundled code with Python and capture output
    // Use Python executable with virtual environment support
    let python_cmd = get_python_executable();

    let python_output = Command::new(&python_cmd)
        .arg(&bundle_path)
        .current_dir(temp_dir.path()) // Set working directory for consistent execution
        .output()
        .map_err(|e| {
            anyhow::anyhow!(
                "Failed to execute Python: {} (command: {} {:?})",
                e,
                python_cmd,
                bundle_path
            )
        })?;

    // Note: For timeout support, we'd need a more complex solution or external crate
    // The standard library doesn't provide built-in timeout for process execution

    // Check for unexpected Python execution failures
    if !python_output.status.success() && !expects_failure {
        let stderr = String::from_utf8_lossy(&python_output.stderr);
        let stdout = String::from_utf8_lossy(&python_output.stdout);

        // This is an unexpected failure - fail the test explicitly
        return Err(anyhow::anyhow!(
            "Python execution failed unexpectedly for fixture '{}':\n\
            Exit code: {}\n\
            Stdout:\n{}\n\
            Stderr:\n{}\n\n\
            If this failure is expected, rename the fixture directory with 'xfail_' prefix.",
            fixture_name,
            python_output.status.code().unwrap_or(-1),
            stdout.trim(),
            stderr.trim()
        ));
    }

    // Create separate snapshots using insta's named snapshot feature
    insta::with_settings!({
        snapshot_suffix => fixture_name,
        omit_expression => true
    }, {
        // Snapshot the bundled code
        insta::assert_snapshot!("bundled_code", bundled_code);

        // Create structured execution results snapshot
        let execution_status = if python_output.status.success() {
            ExecutionStatus::Success
        } else {
            ExecutionStatus::Failed(python_output.status.code().unwrap_or(-1))
        };

        let execution_results = ExecutionResults {
            status: execution_status,
            stdout: String::from_utf8_lossy(&python_output.stdout)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
            stderr: sanitize_paths(
                &String::from_utf8_lossy(&python_output.stderr)
                    .trim()
                    .replace("\r\n", "\n")
            ),
        };

        insta::assert_debug_snapshot!("execution_results", execution_results);

        // Snapshot ruff linting results for cross-validation
        insta::assert_debug_snapshot!("ruff_lint_results", ruff_results);
    });

    Ok(())
}
