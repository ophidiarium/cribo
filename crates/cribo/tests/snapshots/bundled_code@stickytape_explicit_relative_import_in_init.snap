---
source: crates/cribo/tests/test_bundling_snapshots.rs
input_file: crates/cribo/tests/fixtures/stickytape_explicit_relative_import_in_init/main.py
---
#!/usr/bin/env python3
# Generated by Cribo - Python Source Bundler
# https://github.com/ophidiarium/cribo

import types
message = "Hello"
greetings = types.SimpleNamespace(__all__=['message'])
print(message)
